labman/db/process.py

```diff
@@ -3293,14 +3336,37 @@ def generate_prep_information(self):
                     result[nk] = val
 
                 # format some final fields
-                result['platform'] = equipment[
-                    result.pop('platform_id')]['description']
+                result['platform'] = 'Illumina'
+                result['instrument_model'] = ''
+                if assay == self._amplicon_assay_type:
+                    result['extraction_robot'] = '%s_%s' % (
+                        result.pop('epmotion_robot'),
+                        result.pop('kingfisher_robot'))
+                    result['well_description'] = '%s_%s_%s' % (
+                        result['sample_plate'], sid, result['well_id'])
+                    result['primer_plate'] = result[
+                        'primer_plate'].split(' ')[-1]
+                    mgps = marker_gene_primer_set[result.pop('primer_set_id')]
+                    result['PRIMER'] = '%s%s' % (
+                        mgps['linker_sequence'], mgps['fwd_primer_sequence'])
+                    result['pcr_primers'] = 'FWD:%s; REV:%s' % (
+                        mgps['fwd_primer_sequence'],
+                        mgps['rev_primer_sequence'])
+                    result['linker'] = mgps['linker_sequence']
+                    result['target_gene'] = mgps['target_gene']
+                    result['target_subfragment'] = mgps['target_subfragment']
+                    result['library_construction_protocol'] = (
+                        'Illumina EMP protocol {0} amplification of {1}'
+                        ' {2}'.format(mgps['region'], mgps['target_gene'],
+                                      mgps['target_subfragment']))
 
                 if sid is not None and study_id is not None:
                     study = Study(study_id)
                     if study not in data:
                         data[study] = {}
-                    data[study][content] = result
+                    # if we want the sample_name.well_id, just replace sid
+                    # for content
+                    data[study][sid] = result
```

Comment:
I also checked this one with Gail.  She confirms that:
1) The sample_name column must exactly match what is in Qiita (so we can't use sample content there, even if there are duplicates) and
2) Every sample that is in the run must be represented in the prep info file (so, even if there are duplicates, we can't leave those out of the file)
        * Note that while duplicate samples on the same plate or even in the same run is a very undesirable case in the real world, *it is the norm in the test data*.  If you want to change the test db population and attendant unit tests so that they represent a more realistic case (no duplicate samples), that would be great!

In the case of duplicate samples in the same run, Gail manually goes into Qiita and adds a new sample (to represent the duplicate) and then manually modifies the name of the duplicate in the prep file to match the name she created for the new sample in Qiita.  She has asked for Labman to show a warning when users download a prep info file that has duplicate samples in it, telling them that they will need to go through this process.  I have added this as #343 Warn user if prep info file contains duplicate samples.

Reply:
Kewl, just note that in the current state we will have a difference in the blanks and spikes in the prep and the sample information file, which we will need to add to the sample info in Qiita somehow.

Reply:
@antgonza You are absolutely right that the foregoing applies only to experimental samples--thanks for pointing that out!  The known-good example we have for the prep info file shows that blanks are qualified by their well id while samples are not:

![screen shot 2018-08-28 at 10 49 49 am](https://user-images.githubusercontent.com/10677935/44740831-40078200-aab0-11e8-9eae-b1c1140605aa.png)

Furthermore, the `labman.sample_composition` table doesn't even store a `sample_id` for blanks and other controls:

![screen shot 2018-08-28 at 10 52 03 am](https://user-images.githubusercontent.com/10677935/44740891-6d543000-aab0-11e8-9a62-91fba1378bad.png)

Therefore, it looks like the `sample_name` column in the prep info file will need to contain the `sample_id` for experimental samples (i.e., those with `sample_composition_type_id`=1) and the `content` for everything else.  Good catch!

Reply:
Related to @antgonza 's point above about blanks, etc: it looks like handling them requires a modification to our way of generating the Well_Description, too.  The known-good file we got from the lab shows well descriptions like this:

![screen shot 2018-08-28 at 11 04 35 am](https://user-images.githubusercontent.com/10677935/44741634-82ca5980-aab2-11e8-80c8-2aab63c3617c.png)

However, since as noted above the `sample_composition` table doesn't even store a `sample_id` for blanks and controls, following our approach of building the well description from the `sample_id` in these cases gives well descriptions with `None` in them, like this:

![screen shot 2018-08-28 at 11 05 05 am](https://user-images.githubusercontent.com/10677935/44741681-aa212680-aab2-11e8-9286-20609a02bc42.png)

I'm coming to the conclusion that maybe *everywhere* in the prep info file generation that we incorporate `sample_id` for experimental samples (e.g., in sample name, well description, original name, more?), we need to instead incorporate `content` for the non-experimental samples to match the known-good.   @antgonza , what do you think?

Reply:
I think this should be a case by case scenario:
- orig_name = name from qiita or empty for blanks/spikes
- sample_name = original name for unique samples, if samples are duplicated, use content; for blanks/spikes use content
- rest, like well description: use the previous sample name

This is an easy change in the current code; however, not easy to test as we will need to create a test with only unique sample ids and I have _no_ idea of how to do that ...

Reply:
From the known-good we got from the lab, it looks like they currently expect orig_name to match sample_name even for blanks, etc (see below) so let's follow their convention on that.  I agree with the suggestions for all the rest!

![screen shot 2018-08-28 at 11 28 29 am](https://user-images.githubusercontent.com/10677935/44742879-de4a1680-aab5-11e8-9d5f-886063a9d9d9.png)

Re testing: Yeah, testing is going to be sticky :(  The labman test db only knows about ONE Qiita study ("Identification of the Microbiomes for Cannabis Soils") and says that study only has 27 samples (not enough to make a full plate without either repetition or lots of blanks).  It looks like this info comes from qiita/qiita_db/support_files/populate_test_db.sql , rather than from anything within labman, so I'm hesitant to try to modify its content in the labman db population scripts.   I don't want to change the existing test plates in the db to not have any duplicate samples because we need to be able to test the behavior of full plates.  Alternatively, adding new test plates that don't have duplicate samples on them would be a big effort in expanding the db population script; that effort should really be put in at some point, but it is going to be enough of a pain that I don't want to do it until such time as we have access to a larger test study in the qiita tables to "plate".

Given all this, I think the best of a host of bad options may be to change the db population script so that there are a *small number* of unique samples on the existing plates.  This will be a pain and kind of hacky (and will require some changes to the existing unit tests--ugh!) but will mean that we will be able to test all the use cases (unique experimental samples, duplicated experimental samples, and blanks/etc) at least once.  I can probably get a PR ready for that today.

Reply:
Note that it's pretty easy to create new studies in qiita via the qiita-api or even adding new samples to an existing study without having to modify qiita itself but this will need to be done via the qiita_client (which is part of of the [db reset for tests](https://github.com/jdereus/labman/blob/master/labman/db/testing.py#L36-L39)).

Reply:
A PR that puts one unique sample on each of the sample plates in the test db (replacing one of the samples that was previously a repeat) is now available: #346 Test database extension: some unique samples